src/core/utilities/box/box.h

```diff
@@ -2,119 +2,280 @@
 #ifndef PHARE_CORE_UTILITIES_BOX_BOX_H
 #define PHARE_CORE_UTILITIES_BOX_BOX_H
 
-#include <cstddef>
 
 #include "core/utilities/point/point.h"
 #include "core/utilities/meta/meta_utilities.h"
 
+#include <cstddef>
+#include <algorithm>
+#include <optional>
+#include <iostream>
 
-namespace PHARE
+namespace PHARE::core
 {
-namespace core
+template<typename Type, std::size_t dim>
+class box_iterator;
+
+
+
+/** Represents a 1D, 2D or 3D box of integer or floating point
+ * points.
+ */
+template<typename Type, std::size_t dim>
+struct Box
 {
-    /** Represents a 1D, 2D or 3D box of integer or floating point
-     * points.
-     */
-    template<typename Type, std::size_t dim>
-    struct Box
+    Point<Type, dim> lower;
+    Point<Type, dim> upper;
+    std::array<Type, dim> shape;
```

Comment:
this needs to be kept in sync with the lower upper if they ever change, probably better to just use the shape() that was merged in #629